### PR TITLE
`Communication`: Accept messaging code of conduct

### PR DESCRIPTION
**Pull request by nityanandaz**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
https://github.com/ls1intum/Artemis/issues/7085: Accept a course's code of conduct

### Description
Add a guard such that a user can only write messages if they accepted the course's code of conduct.

### Steps for Testing

Prerequisites:
- 1 Student/Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Messages
3. Accept the Code of Conduct
4. Click the _Code of Conduct_ button to look it up
5. Navigate elsewhere
6. Navigate to Messages, the Code of Conduct should remain accepted

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [user.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/core/user/user.model.ts.html) | 100% | ✅ |
| [course-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-update.component.ts.html) | 82.3% | ✅ |
| [course-conversations-code-of-conduct.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.ts.html) | 92.85% | ✅ |
| [course-conversations.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/course-conversations.component.ts.html) | 100% | ✅ |
| [conversation-selection-sidebar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.ts.html) | 93.25% | ✅ |
| [file.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/http/file.service.ts.html) | 50% | ✅ |
| [conversation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/conversations/conversation.service.ts.html) | 96.38% | ✅ |
| [metis-conversation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis-conversation.service.ts.html) | 83.79% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ConductAgreement.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/ConductAgreement.html) | 62% | ✅ |
| [ConductAgreementId.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/ConductAgreementId.html) | 43% | ✅ |
| ConductAgreementRepository.java | not found (added) | ❌ |
| [ConductAgreementService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ConductAgreementService.html) | 100% | ✅ |
| [ResponsibleUserDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.dto/ResponsibleUserDTO.html) | 100% | ✅ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅ |
| [FileResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/FileResource.html) | 72% | ✅ |
| [ConversationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5385/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ConversationResource.html) | 92% | ✅ |

### Screenshots

image
image

https://github.com/ls1intum/Artemis/assets/23535000/8719190a-5c62-4746-b647-9d55fde9579a

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/metis/ConversationIntegrationTest.java [hunk 1]
```diff
@@ -428,6 +428,21 @@
...
+    void codeOfConduct_isAccepted() throws Exception {
+        var agreement = request.get("/api/courses/" + exampleCourseId + "/code-of-conduct-agreement", HttpStatus.OK, Boolean.class);
+        assertThat(agreement).isFalse();
+    }
+
+    @Test
+    @WithMockUser(username = TEST_PREFIX + "student1", roles = "USER")
+    void codeOfConduct_accept() throws Exception {
+        var agreement = request.postWithResponseBody("/api/courses/" + exampleCourseId + "/code-of-conduct-agreement", HttpStatus.OK, Boolean.class);
+        assertThat(agreement).isTrue();
```
Comment (TimOrtel) [inline src/test/java/de/tum/in/www1/artemis/metis/ConversationIntegrationTest.java:442]: I feel like this test is too weak. Can you please add an additional verification that a subsequent call to get the status of the agreement now returns true. 

```java
var agreement = request.postWithResponseBody("/api/courses/" + exampleCourseId + "/code-of-conduct-agreement", HttpStatus.OK, Boolean.class);
        assertThat(agreement).isTrue();

        var newAgreement = request.get("/api/courses/" + exampleCourseId + "/code-of-conduct-agreement", HttpStatus.OK, Boolean.class);
        assertThat(newAgreement).isTrue();
```

File: src/test/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    @Autowired
+    private UserUtilService userUtilService;
+
+    @BeforeEach
+    void init() {
+        userUtilService.addUsers(TEST_PREFIX, 1, 0, 0, 3);
+    }
+
+    @Test
+    void fetchAndAgreeIsCodeOfConductAccepted() {
```
Comment (TimOrtel) [inline src/test/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementServiceTest.java:40]: Good test 👍

File: src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * A user agrees to a course's code of conduct.
+     *
+     * @param user   the user in the course
+     * @param course the code of conduct's course
+     */
+    public void setUserAgreesToCodeOfConductInCourse(User user, Course course) {
+        CourseCodeOfConductAgreement courseCodeOfConductAgreement = courseCodeOfConductAgreementRepository.findByCourseIdAndUserId(course.getId(), user.getId())
+                .orElse(new CourseCodeOfConductAgreement());
+        courseCodeOfConductAgreement.setCourse(course);
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java:55]: Why is this necessary? Would it not be possible to always start with `new CourseOfConductAgreement()`?

File: src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+    <changeSet author="zbil" id="20230908050001">
+        <createTable tableName="course_code_of_conduct_agreement">
```
Comment (TimOrtel) [inline src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml:6]: I think there are multiple foreign keys missing here, however I am not a liquibase notation expert (have you maybe defined them somewhere else?).
1. On Delete Cascade Foreign Key From course_id to table courses.
2. On Delete Cascade Foreign Key from user_id to table users.

File: src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        let label = '';
+        if (user.firstName) {
+            label += user.firstName + ' ';
+        }
+        if (user.lastName) {
+            label += user.lastName + ' ';
+        }
+        return label.trim();
+    }
+}
```
Comment (TimOrtel) [inline src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.ts:47]: I feel like there should already be a public helper method to derive the human-readable name for a user. Have you searched for one?

File: src/main/webapp/app/shared/metis/metis-conversation.service.ts [hunk 6]
```diff
@@ -102,7 +112,16 @@ export class MetisConversationService implements OnDestroy {
         }
         this.activeConversation = cachedConversation;
         this._activeConversation$.next(this.activeConversation);
-    };
+        this.isCodeOfConductPresented = false;
+        this._isCodeOfConductPresented$.next(this.isCodeOfConductPresented);
+    }
+
+    public setCodeOfConduct() {
```
Comment (TimOrtel) [inline src/main/webapp/app/shared/metis/metis-conversation.service.ts:119]: What does this do? Can you add a documentation to this public method?

**Review by TimOrtel — COMMENTED** [review]
Thank you for this PR. I like you code style, it looks very tidy and cleaned up. I left some comments but nothing blocking.

File: src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    @NotNull
+    private Course course;
+
+    @ManyToOne
+    @JsonIncludeProperties({ "id" })
+    @NotNull
+    private User user;
+
+    @Column(name = "is_accepted")
+    private Boolean isAccepted;
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java:31]: Can you make this a primitive boolean and adjust the getter and setter correspondingly? I think you can also initialize the field with false.

File: src/main/java/de/tum/in/www1/artemis/service/dto/UserPublicInfoDTO.java [hunk 8]
```diff
@@ -25,6 +25,8 @@ public class UserPublicInfoDTO {
 
     private String lastName;
 
+    private String email;
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/dto/UserPublicInfoDTO.java:28]: Are you sure this is a good idea? If I read this correctly, this will expose the emails of all users of the course to a widely accessible endpoint. Maybe you should apply an additional filtering that only sets this field for instructs, but not for tutors and students.

File: src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+            <column autoIncrement="true" name="id" type="bigint">
+                <constraints nullable="false" primaryKey="true"/>
+            </column>
+            <column name="course_id" type="bigint"/>
+            <column name="user_id" type="bigint"/>
+            <column name="is_accepted" type="boolean"/>
```
Comment (lennart-keller) [inline src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml:12]: Maybe you can also add a constraint with nullable="false" and also set a default value of false.

**Comment by TimOrtel** [comment]
Currently, when you change the code of conduct, users are not presented with the updated code of conduct again. I think, when editing the code of conduct, there should be a checkbox that allows instructors to force all users to re-accept the code of conduct.

**Comment by TimOrtel** [comment]
Currently, the API does not decline API requests for the conversation feature (for example a user creating a new message) when they have not yet accepted the code of conduct.

File: src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+     * @param courseId the ID of the code of conduct's course
+     * @param userId   the user's ID
+     * @return the user's agreement to the course's code of conduct
+     */
+    @Query("""
+            SELECT c
+            FROM CourseCodeOfConductAgreement c
+            WHERE c.course.id = :courseId AND c.user.id = :userId
+            """)
+    Optional<CourseCodeOfConductAgreement> findByCourseIdAndUserId(@Param("courseId") Long courseId, @Param("userId") Long userId);
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java:30]: Optional: I think you can even leave the query out and let spa generate it itself using the method name.

File: src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        if (courseCodeOfConductAgreement.isPresent()) {
+            return courseCodeOfConductAgreement.get().getIsAccepted();
+        }
+        else {
+            CourseCodeOfConductAgreement courseCodeOfConductAgreementNew = new CourseCodeOfConductAgreement();
+            courseCodeOfConductAgreementNew.setCourse(course);
+            courseCodeOfConductAgreementNew.setUser(user);
+            courseCodeOfConductAgreementNew.setIsAccepted(false);
+            courseCodeOfConductAgreementRepository.save(courseCodeOfConductAgreementNew);
+            return courseCodeOfConductAgreementNew.getIsAccepted();
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java:42]:
```suggestion
            return false;
```
This makes the returned value much clearer

File: src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java [hunk 12]
```diff
@@ -124,6 +129,39 @@ public ResponseEntity<Boolean> hasUnreadMessages(@PathVariable Long courseId) {
...
+     */
+    @PostMapping("/{courseId}/code-of-conduct-agreement")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Boolean> acceptCodeOfConduct(@PathVariable Long courseId) {
+        checkMessagingEnabledElseThrow(courseId);
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var requestingUser = userRepository.getUserWithGroupsAndAuthorities();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, courseRepository.findByIdElseThrow(courseId), requestingUser);
+        courseCodeOfConductAgreementService.setUserAgreesToCodeOfConductInCourse(requestingUser, course);
+        return ResponseEntity.ok(courseCodeOfConductAgreementService.fetchUserAgreesToCodeOfConductInCourse(requestingUser, course));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java:162]: Why do you have to fetch it again from the data base? And why do you need the returned value at all, shouldn't it be always true?

File: src/main/resources/templates/codeofconduct/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+- The use of sexualized language or imagery, and sexual attention or advances of any kind
+- Trolling, insulting or derogatory comments, and personal or political attacks
+- Public or private harassment
+- Publishing others' private information, such as a physical or email address, without their explicit permission
+- Other conduct which could reasonably be considered inappropriate in a professional setting
+
+## Scope
+
+This Code of Conduct applies within all communication channels.
```
Comment (JohannesStoehr) [inline src/main/resources/templates/codeofconduct/README.md:27]: Maybe also include everything else in the course, e.g. submissions, in person communication, ...

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/in/www1/artemis/metis/ConversationIntegrationTest.java (hunk 1 above)
Comment (nityanandaz) [reply src/test/java/de/tum/in/www1/artemis/metis/ConversationIntegrationTest.java:442]: Agree c1da199fdaf91b5cee8e27664664a154d4f46033

File: src/test/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementServiceTest.java (hunk 2 above)
Comment (nityanandaz) [reply src/test/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementServiceTest.java:40]: Thanks!

File: src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java (hunk 7 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java]: Agree 59b450f4a596ec5d8e9361b399de2d32a1d5ec48

File: src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml (hunk 9 above)
Comment (nityanandaz) [reply src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml]: I revised it. I assume BIT cannot be NULL :) 1a7e7089c96d9cddb5cb328e82e3f0a5a25a500e

File: src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java (hunk 3 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java]: Agree 45fbbc7319e6e8bb6df676f8b676671e698f4896

File: src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.ts (hunk 5 above)
Comment (nityanandaz) [reply src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.ts]: I use it now 45fbbc7319e6e8bb6df676f8b676671e698f4896

File: src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java (hunk 10 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java]: I didn't know, thanks! 45fbbc7319e6e8bb6df676f8b676671e698f4896

File: src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java (hunk 11 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/service/CourseCodeOfConductAgreementService.java]: The agreement is now modeled with the presence/absence of the db entry 45fbbc7319e6e8bb6df676f8b676671e698f4896

File: src/main/webapp/i18n/de/courseCodeOfConduct.json [hunk 14]
```diff
@@ -0,0 +1,9 @@
+{
+    "artemisApp": {
+        "codeOfConduct": {
+            "accept": "Akzeptieren",
+            "title": "Verhaltenskodex",
+            "tooltip": "Der Verhaltenskodex gibt Nutzer:innen an, wie sie miteinander kommunizieren sollen und welche Konsequenzen bei Fehlverhalten drohen können, sowie einen Kontakt zur Berichterstattung."
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/webapp/i18n/de/courseCodeOfConduct.json]: Thanks 4d662bfa383f0fbe620d3c998b95ce17c738cbac

File: src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml (hunk 4 above)
Comment (nityanandaz) [reply src/main/resources/config/liquibase/changelog/20230908050001_changelog.xml]: The new migration adds a foreign key constraint

File: src/main/resources/templates/codeofconduct/README.md (hunk 13 above)
Comment (nityanandaz) [reply src/main/resources/templates/codeofconduct/README.md:27]: From a technical perspective, the code of conduct only guards the Messages feature

File: src/main/webapp/app/shared/metis/metis-conversation.service.ts (hunk 6 above)
Comment (nityanandaz) [reply src/main/webapp/app/shared/metis/metis-conversation.service.ts:122]: It shows the code of conduct baeeaf3e73d143a98619b323696bcf092d709b20

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me now

File: src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+            return false;
+        if (!super.equals(o))
+            return false;
+        CourseCodeOfConductAgreement that = (CourseCodeOfConductAgreement) o;
+        return course.equals(that.course) && user.equals(that.user);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), course, user);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java:58]:
```suggestion
        return Objects.hash(course, user);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 16]
```diff
@@ -304,6 +304,21 @@ public ResponseEntity<byte[]> getCourseIcon(@PathVariable Long courseId) {
...
+     * @return The requested file, 403 if the logged-in user is not allowed to access it, or 404 if the file doesn't exist
+     */
+    @GetMapping("files/templates/code-of-conduct")
+    @EnforceAtLeastInstructor
+    public ResponseEntity<byte[]> getCourseCodeOfConduct() throws IOException {
+        var templatePath = Path.of("templates", "codeofconduct", "README.md");
+        log.debug("REST request to get template : {}", templatePath);
+        var resource = resourceLoaderService.getResource(templatePath);
+        var path = resource.getFile().getPath();
+        return responseEntityForFilePath(path);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:319]: Please use `resource.getInputStream()` here to load the content, this won't work if the resource is bundled into the application.

File: src/main/resources/templates/codeofconduct/README.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
+# Code of Conduct Template: Adapt to your demands
```
Comment (Strohgelaender) [inline src/main/resources/templates/codeofconduct/README.md:1]: In case the instructor does not change anything, the word "template" should not be shown to the students.
```suggestion
<!-- Code of Conduct Template: Adapt to your demands -->
```

**Review by Strohgelaender — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.html [hunk 18]
```diff
@@ -0,0 +1,8 @@
+<h3>{{ 'artemisApp.codeOfConduct.title' | artemisTranslate }}</h3>
+<div [innerHTML]="course.courseInformationSharingMessagingCodeOfConduct || '' | htmlForMarkdown"></div>
+<ul>
+    <li *ngFor="let user of responsibleContacts">
+        {{ user.name }} (<a href="mailto:{{ user.email }}">{{ user.email }}</a
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/webapp/app/overview/course-conversations/code-of-conduct/course-conversations-code-of-conduct.component.html:5]: Yes, we agreed on mailto:

File: src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java [hunk 19]
```diff
@@ -124,6 +132,61 @@ public ResponseEntity<Boolean> hasUnreadMessages(@PathVariable Long courseId) {
...
+     * @param courseId the course's ID
+     * @return ResponseEntity with status 200 (Ok) and body is true if the user agreed to the course's code of conduct
+     */
+    @GetMapping("/{courseId}/code-of-conduct/agreement")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Boolean> isCodeOfConductAccepted(@PathVariable Long courseId) {
+        checkMessagingEnabledElseThrow(courseId);
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var requestingUser = userRepository.getUserWithGroupsAndAuthorities();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, courseRepository.findByIdElseThrow(courseId), requestingUser);
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java]: An oversight :) 35bf2d4a550ff8d2f8281e11f0c0d9ed101c5dc5

File: src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java [hunk 20]
```diff
@@ -124,6 +132,61 @@ public ResponseEntity<Boolean> hasUnreadMessages(@PathVariable Long courseId) {
...
+     * @param courseId the course's ID
+     * @return ResponseEntity with status 200 (Ok)
+     */
+    @PatchMapping("/{courseId}/code-of-conduct/agreement")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Void> acceptCodeOfConduct(@PathVariable Long courseId) {
+        checkMessagingEnabledElseThrow(courseId);
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var requestingUser = userRepository.getUserWithGroupsAndAuthorities();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, courseRepository.findByIdElseThrow(courseId), requestingUser);
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/web/rest/metis/conversation/ConversationResource.java]: An oversight :) 35bf2d4a550ff8d2f8281e11f0c0d9ed101c5dc5

File: src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java (hunk 15 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/domain/CourseCodeOfConductAgreement.java]: Ok :) 0f80c151ebe75fdc2c3ab02640561b8977893adf

File: src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import de.tum.in.www1.artemis.domain.CourseCodeOfConductAgreement;
+
+/**
+ * Spring Data repository for the Code of Conduct Agreement entity.
+ */
+@Repository
+public interface CourseCodeOfConductAgreementRepository extends JpaRepository<CourseCodeOfConductAgreement, Long> {
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseCodeOfConductAgreementRepository.java]: Thanks 0f80c151ebe75fdc2c3ab02640561b8977893adf

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java (hunk 16 above)
Comment (nityanandaz) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java]: Thanks 0f80c151ebe75fdc2c3ab02640561b8977893adf

File: src/main/resources/config/liquibase/changelog/20230919184052_changelog.xml [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+    </changeSet>
+    <changeSet author="zbil" id="1695141658157-224">
+        <addForeignKeyConstraint baseColumnNames="course_id" baseTableName="course_code_of_conduct_agreement"
+                                 constraintName="FK8sijgspbqroh9saasrpj3ifsa" deferrable="false"
+                                 initiallyDeferred="false" referencedColumnNames="id" referencedTableName="course"
+                                 validate="true"/>
+    </changeSet>
+    <changeSet author="zbil" id="1695141658157-226">
+        <addForeignKeyConstraint baseColumnNames="user_id" baseTableName="course_code_of_conduct_agreement"
+                                 constraintName="FKiuc9i7eratit8nw8od8u6bb5p" deferrable="false"
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply src/main/resources/config/liquibase/changelog/20230919184052_changelog.xml]: Yes :) 0f80c151ebe75fdc2c3ab02640561b8977893adf

File: src/main/resources/templates/codeofconduct/README.md (hunk 17 above)
Comment (nityanandaz) [reply src/main/resources/templates/codeofconduct/README.md]: Agree 0f80c151ebe75fdc2c3ab02640561b8977893adf

**Review by florian-glombik — CHANGES_REQUESTED** [review]
Should be done before a merge:
* **Make a migration to give all existing courses the default code of conduct** _(make sure they also work on postgres)_
* **I do not think that it is a good idea to display all instructors in the Code of Conduct** -> preferred solution: enforce that at least 1 contact person is defined when setting up the Code of Conduct but do **not** display all instructors 
Scenario where it breaks: E.g. some editors are instructors for creating an exam, also a specific instructor might be responsible for organizational concerns such as the Code of Conduct, but not all instructors

Can be done in a follow up:
* Do not show the code of conduct popup if no code of conduct is present _(somebody might remove the text)_
* Also show the code of conduct for communications

**Comment by nityanandaz** [comment]
@florian-glombik we decided on

> * Do not show the code of conduct popup if no code of conduct is present _(somebody might remove the text)_

instead of a migration (ccf73fd9d0a0237f14aa6fc15020fec4f1518d23).

- explicitly setting responsible contacts is considered for a follow-up
- the code of conduct will not cover the Communication tab

**Review by TimOrtel — APPROVED** [review]
Code Looks good to me

**Review by tobias-lippert — APPROVED** [review]
Reapprove after clarifiying that the change requests are no longer valid.

**Review by lennart-keller — APPROVED** [review]
Code lgtm 👍

**Review by florian-glombik — APPROVED** [review]
Re-approve since no major changes since last review https://github.com/ls1intum/Artemis/pull/7154#issuecomment-1754952244 and issues are clarified / resolved in follow-ups
